rl_zoo3/feature_extractors.py

```diff
@@ -0,0 +1,250 @@
+        and bounds (values in [0, 255]).
+    :param out_features_per_frame: Number of extracted features from a single image.
+        This corresponds to the number of unit in the last layer.
+        If the frames are stacked, the output dimension will be equal to
+        out_features_per_frame * stacking_frames.
+    :param linear_activation_fn: The activation function to use after the last layer.
+    :param stacking_frames: The number of the stacked frames.
+        Must be the same as the hyperparamter `frame_stack`.
+    :param frame_channels: The number of color channels in the input images.
+        Typically (1) for monochromatic, (3) for RGB and (4) for RGBD.
+    """
+
+    def __init__(
+        self,
+        observation_space: gym.Space,
+        out_features_per_frame: int,
+        linear_activation_fn: Type[nn.Module],
+        stacking_frames: int | None,
+        frame_channels: int,
+        normalized_image: bool,
+    ):
+        self._validate_params(observation_space, normalized_image, stacking_frames, frame_channels)
+        if self.is_single_frame:
+            stacking_frames = 1
+        output_dim = out_features_per_frame * stacking_frames
+        super().__init__(observation_space, output_dim)
+
+        feature_extractor = self._prepare_feature_extractor()
+        flatten = nn.Flatten()
+
+        # Compute shape by doing one forward pass
+        with th.no_grad():
+            obs = th.as_tensor(observation_space.sample()[None]).float()
+            if not self.is_single_frame:
+                obs = th.split(obs, frame_channels, dim=1)[0]
+            n_flatten = flatten(feature_extractor(obs)).squeeze().shape[0]
+
+        # The output dimension of the linear layer is just for one frame.
+        # The result will match the output_dim in the forward() method.
+        linear = nn.Sequential(nn.Linear(n_flatten, out_features_per_frame), linear_activation_fn())
+        self.fe_model = nn.Sequential(feature_extractor, flatten, linear)
+
+    def _validate_params(
+        self, observation_space: gym.Space, normalized_image: bool, stacking_frames: int | None, frame_channels: int
+    ):
+        assert isinstance(observation_space, spaces.Box), (
+            "PreTrainedVisionExtractor must be used with a gym.spaces.Box ",
+            f"observation space, not {observation_space}",
+        )
+        self.is_single_frame = stacking_frames is None or stacking_frames < 2
+        if self.is_single_frame:
+            assert is_image_space(observation_space, check_channels=True, normalized_image=normalized_image), (
+                f"You should use a VisionExtractor only with images not with {observation_space}.\n"
+                "If the `stackig_frames` is greater than 1, please set the Extractor params `stackig_frames`and `frame_channels` accordingly."
+            )
+        else:
+            self.frame_channels = frame_channels
+            n_channels = observation_space.shape[0]
```

Comment:
Maybe instead of two lines use `numpy.divmod` function?
Usage:
```python
divisions, modulo = np.divmod(17, 4)  # (4, 1)
```

Reply:
(i meant lines 90 and 91 but i've inserted the comment poorly)

Reply:
Applied simplification, thanks, i forget about thus one liner